xml_loader/shift_generation.py

```diff
@@ -0,0 +1,261 @@
+        for e in emp:
+            if(c in e.competencies):
+                employee_with_competencies[c].append(int(e.id))
+    
+    for e in emp:
+        id = int(e.id)
+        employees.append(id)
+        employee_daily_rest[id] = e.daily_rest_hours
+        employee_weekly_rest[id] = e.daily_rest_hours
+        employee_contracted_hours[id] = e.contracted_hours
+        
+    return employees, employee_with_competencies, employee_weekly_rest, employee_daily_rest, employee_contracted_hours
+
+
+def get_durations(root):
+    events = get_events(root)
+    durations = {}
+    #Possible durations are between 6.0 hours and 12.0 hours
+    possible_durations = [t/4 for t in range(6*4, (12*4+1))]
+    for t in events:
+        for dur in possible_durations:
+            if(t+dur in events):
+                try:
+                    durations[t].append(dur)
+                except:
+                    durations[t] = [dur]
+    return durations
+
+def get_shift_lists(root):
+    durations = get_durations(root)
+    shifts_per_day = tupledict()
+    shifts = tuplelist()
+    days = get_days(root)
+    time_step = get_time_steps(root)
+    for d in days:
+        shifts_per_day[d] = []
+        for t in durations:
+            if(t >=d*24 and t <= (24*(d+1) - time_step)):
+                for dur in durations[t]:
+                    shifts_per_day[d].append((t,dur))
+                    shifts.append((t,dur))
+            if(t > 24*d):
+                continue
+    return [shifts, shifts_per_day]
+
+def get_shift_list(root):
+    shifts = tuplelist()
+    dur = get_durations(root)
+    i = 0
+    for t in dur:
+        for v in dur[t]:
+            shifts.append(i)
+            i += 1
+    return shifts
+
+def get_shifts_overlapping_t(root):
+    time_periods = get_time_periods(root)[0]
+    time_step = get_time_steps(root)
+    shifts_overlapping_t = {}
+    shifts = get_durations(root)
+    for t in time_periods:
+        for time in shifts:
+            for dur in shifts[time]:
+                if(t >= time and t < time + dur):
+                    try:
+                        shifts_overlapping_t[t].append((time, dur))
+                    except:
+                        shifts_overlapping_t[t] = [(time, dur)]
+    return shifts_overlapping_t
+
+def get_start_events(root):
+    events = []
+    demand_days = get_days_with_demand(root)
+    for day in demand_days:
+        for t in range(len(demand_days[day].start)):
+            for j in range(len(demand_days[day].end)):
+                diff = demand_days[day].end[j] - demand_days[day].start[t]
+                if(diff >= 6):
+                    events.append((demand_days[day].start[t] + 24*day))
+                    break
+    return events
+
+
+def get_off_shifts(root):
+    events = get_start_events(root)
+    off_shifts = []
+    off_shifts_in_week = tupledict()
+    week = 0
+    off_shifts_in_week[week] = []
+    for i in range(len(events)):
+        for event2 in events[i:]:
```

Comment:
Hvorfor ikke `event` i stedet for `event2`?

Reply:
Husker ikke. Den kan nok godt endres til event